falcon/cluster/cluster.py

```diff
@@ -2,622 +2,341 @@
-    else:
-        index = faiss.IndexIVFFlat(
-            faiss.IndexFlatIP(dim), dim, n_list, faiss.METRIC_INNER_PRODUCT
-        )
-        index.nprobe = min(math.ceil(index.nlist / 8), n_probe)
-    # Compute cluster centroids.
-    # noinspection PyArgumentList
-    index.train(train_vectors)
-
-    # Add the vectors to the index in batches.
-    batch_start = 0
-    for batch in tqdm(
-        dataset.to_batches(max_chunksize=batch_size),
-        desc="Batches added to index",
-        unit="batch",
-    ):
-        batch_spectra = batch.to_pandas().apply(
-            spectrum.df_row_to_spec, axis=1
-        )
-        batch_vectors = []
-        for spec in batch_spectra:
-            batch_vectors.append(spec.vector)
-            precursor_mzs.append(spec.precursor_mz)
-            rts.append(spec.retention_time)
-        batch_stop = batch_start + len(batch_vectors)
-        vectors.extend(batch_vectors)
-
-        batch_vectors = np.stack(batch_vectors)
-        index.add_with_ids(
-            batch_vectors,
-            np.arange(batch_start, batch_stop, dtype=np.int64),
-        )
-        batch_start = batch_stop
-    precursor_mzs = np.asarray(precursor_mzs)
-    rts = np.asarray(rts)
-    # Query the index to calculate NN distances.
-    _dist_mz_interval(
-        index,
-        vectors,
-        precursor_mzs,
-        rts,
-        batch_size,
-        n_neighbors,
-        precursor_tol_mass,
-        precursor_tol_mode,
-        rt_tol,
-        distances,
-        indices,
-        indptr,
+    # Hierarchical clustering using the precomputed pairwise distance matrix.
+    min_samples = 2
+    logger.debug(
+        "Hierarchical clustering (distance_threshold=%.4f, min_samples=%d)",
+        distance_threshold,
+        min_samples,
     )
-    index.reset()
-    indptr_i += n_spectra
-
-    metadata = (
+    # Sort the metadata by increasing precursor m/z for easy subsetting.
+    data = (
         dataset.to_table(
             columns=[
-                "filename",
-                "identifier",
                 "precursor_mz",
+                "precursor_charge",
                 "retention_time",
-            ],
-            scan_in_order=True,
+                "mz",
+                "intensity",
+            ]
         )
+        .add_column(0, "index", pa.array(range(dataset.count_rows())))
+        .sort_by("precursor_mz")
         .to_pandas()
-        .rename(columns={"identifier": "spectrum_id"})
     )
-    return metadata
+    # Cluster per contiguous block of precursor m/z's (relative to the
+    # precursor m/z threshold).
+    logging.info(
+        "Cluster %d spectra using %s linkage and distance threshold %.3f",
+        len(data),
+        linkage,
+        distance_threshold,
+    )
+    with tempfile.NamedTemporaryFile(suffix=".npy") as cluster_file:
+        cluster_filename = cluster_file.name
+    cluster_labels = np.lib.format.open_memmap(
```

Comment:
**todo:** I think this code should be within the `with` block, because after you exit it, the temporary file is not guaranteed to keep existing but can be cleaned up by the OS. Alternatively, don't use a context manager.